### DISCUSSION TITLE
How to set trial params from optuna.distibution?

### DISCUSSION THREAD

**Discussion by GooseIt** [optuna/optuna#4554] [Q&A] [answered]
I am using `optuna` for function optimization.

I have `params_dict`, specifying names of all parameters that could be tuned, as well as their `optuna.distribution` objects, as example:

    params_dict = {
        "n_iterations": optuna.distributions.CategoricalDistribution((10, 100, 200)),
        "learning_rate": optuna.distributions.LogUniformDistribution(1e-4, 1e-2)
    }

I have to optimize the function of the parameters using `optuna.study.optimize`.

In it, parameters can be set using `trial.suggest_*` (as example, `trial.suggest_uniform`) functions. The problem is that there is no way to suggest trial params from `optuna.distribution` object directly (the corresponding `trial.suggest` function was made deprecated).

If I to use proposed `trial.suggest_*` functions, my code will have to use multiple if clauses to set trial params. As example, for a given `param_name` with `param_distribution` I will have to go through all the possible distributions of said param:

    if type(param_distribution) == optuna.distributions.UniformDistribution:
        params[param_name] = trial.suggest_uniform(...)
    elif type(param_distribution) == optuna.distributions.CategoricalDistribution:
        params[param_name] = trial.suggest_categorical(...)
    ...

Is there an elegant and lightweight solution that allows to set `trial.params` directly from `params_dict` described above?

**Comment by nzw0301** [comment, marked as answer]
How about defined-and-run with the ask-and-tell interface? The example is available at https://optuna.readthedocs.io/en/stable/tutorial/20_recipes/009_ask_and_tell.html#define-and-run.
